### PR TITLE
fix: Filter out ALL_PERMISSIONS from keys in LSP6_DEFAULT_PERMISSIONS.

### PR DESCRIPTION
**Pull request by richtera**
### What kind of change does this PR introduce (bug fix, feature, docs update, ...)?

Filter out keys from LSP6_DEFAULT_PERMISSIONS because it now has ALL_PERMISSIONS.
Having ALL_PERMISSIONS set during encodePermissions will always return ALL_PERMISSIONS.
Currently calling decodePermissions will always return ALL_PERMISSIONS on.

### CODE REVIEW

File: .vscode/settings.json
Comment (CJ42) [inline .vscode/settings.json:1]: do we want to have these files?

File: src/lib/permissions.ts [hunk 1]
```diff
@@ -1,60 +1,37 @@
...
-  // Do not use an for of loop here to not require the regenerator runtime
-  for (let i = 0; i < additionalPermissions.length; i += 1) {
-    const permission = additionalPermissions[i];
-    if (permissions[permission]) {
-      basePermissions |= BigInt(
-        hexToNumber(LSP6_DEFAULT_PERMISSIONS[permission]),
-      );
-    }
-  }
+  let basePermissions = BigInt(0)
```
Comment (richtera) [inline src/lib/permissions.ts:6]: This is where the encodePermissions was changed.

File: src/lib/permissions.ts [hunk 2]
```diff
@@ -83,33 +60,27 @@ export function decodePermissions(permissionHex: string) {
     SIGN: false,
```
Comment (richtera) [inline src/lib/permissions.ts:60]: This is where the decodePermissions was changed.

**Review by CJ42 — CHANGES_REQUESTED** [review]
can we remove biome from this PR, and put it in a separate PR please?
Otherwise the PR is really hard to review as there are too many changes and I can't filter out what is related to the permissions

File: src/index.test.ts [hunk 3]
```diff
@@ -1382,170 +1393,223 @@ describe('Running @erc725/erc725.js tests...', () => {
...
           encoded,
           LSP6_DEFAULT_PERMISSIONS.ALL_PERMISSIONS,
-          'Encoded permissions do not match expected value for ALL_PERMISSIONS',
-        );
-      });
+          'Encoded permissions do not match expected value for ALL_PERMISSIONS'
+        )
+      })
+
+      it('should decode ALL_PERMISSIONS', () => {
```
Comment (richtera) [inline src/index.test.ts:1517]: Here is the first new test

File: src/index.test.ts [hunk 4]
```diff
@@ -1382,170 +1393,223 @@ describe('Running @erc725/erc725.js tests...', () => {
...
+        const decodedPermissions = decodePermissions(encoded)
+
+        assert.strictEqual(
+          decodedPermissions.ALL_PERMISSIONS,
+          true,
+          'Decoded permissions includes ALL_PERMISSIONS'
+        )
+      })
+
+      it('should not decode CALL or ALL_PERMISSIONS if perms are missing', () => {
```
Comment (richtera) [inline src/index.test.ts:1530]: Here is the second new test

**Review by richtera — COMMENTED** [review]
Since there are a lot of style changes here are the real changes.

File: .vscode/settings.json
Comment (richtera) [reply .vscode/settings.json:1]: Yes, these files are necessary to make sure any devs use the correct settings. It's kind of unfortunate but currently .editorconfig doesn't properly defined formatters and validators in a generic way. If this were to use a standard lsp https://microsoft.github.io/language-server-protocol/ then any editor could hook into it.

**Comment by richtera** [comment]
I removed import sorting, so now only the cleanup and fixes are there.

File: package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@erc725/erc725.js",
-  "version": "0.24.0",
+  "version": "0.24.1-dev.5",
```
Comment (CJ42) [inline package.json:3]: why `-dev.5`?

File: package.json [hunk 6]
```diff
@@ -16,6 +16,8 @@
     "build:module": "tsc -p tsconfig.module.json",
     "test": "env TS_NODE_COMPILER_OPTIONS='{\"module\": \"commonjs\" }' nyc --reporter=text --reporter=lcov mocha",
     "lint": "eslint . --ext .ts",
+    "format:fix": "prettier --write .",
```
Comment (CJ42) [inline package.json:19]: nice addition 👌

File: src/index.test.ts [hunk 7]
```diff
@@ -1498,12 +1505,81 @@ describe('Running @erc725/erc725.js tests...', () => {
...
+
+        assert.strictEqual(
+          decodedPermissions.ALL_PERMISSIONS,
+          true,
+          'Decoded permissions includes ALL_PERMISSIONS',
+        );
+      });
+
+      it('should not decode CALL or ALL_PERMISSIONS if perms are missing', () => {
+        const permissions = { ALL_PERMISSIONS: true, CALL: false };
```
Comment (CJ42) [inline src/index.test.ts:1522]: @skimaharvey was not the intended logic to be "all permissions except `CALL`" in this kind of case?

File: src/index.test.ts [hunk 8]
```diff
@@ -1498,12 +1505,81 @@ describe('Running @erc725/erc725.js tests...', () => {
...
+        const decodedPermissions = decodePermissions(encoded);
+
+        assert.strictEqual(
+          decodedPermissions.CALL,
+          false,
+          'Decoded permissions includes CALL',
+        );
+        assert.strictEqual(
+          decodedPermissions.ALL_PERMISSIONS,
+          false,
```
Comment (CJ42) [inline src/index.test.ts:1534]: shouldn't this be `true`?

```suggestion
          true,
```

File: src/index.test.ts [hunk 9]
```diff
@@ -1498,12 +1505,81 @@ describe('Running @erc725/erc725.js tests...', () => {
...
+
+        assert.strictEqual(
+          redecodedPermissions.CALL,
+          false,
+          'Re-reencoded permissions includes CALL',
+        );
+      });
+
+      it('should allow editing of permissions add extra', () => {
+        const permissions = { ALL_PERMISSIONS: true };
```
Comment (CJ42) [inline src/index.test.ts:1556]: how about a test like that:

```suggestion
        const permissions = { ALL_PERMISSIONS: true, SUPER_DELEGATECALL: true, DELEGATECALL: true };
```

Should lead to the same result right?

File: src/lib/permissions.ts [hunk 10]
```diff
@@ -87,22 +59,16 @@ export function decodePermissions(permissionHex: string) {
...
-      result[testPermission] = true;
-    }
-    return result;
-  }
-
-  const passedPermissionDecimal = Number(hexToNumber(permissionHex));
+  const passedPermissionDecimal = BigInt(hexToNumber(permissionHex, true));
 
   // Do not use an for of loop here to not require the regenerator runtime
+  // Deal with ALL_PERMISSIONS the same way. So as long as all the bits in ALL_PERMISSIONS
```
Comment (CJ42) [inline src/lib/permissions.ts:65]: I am very confused by this comment. I think might need a bit of re-writing and maybe a small example?

**Review by CJ42 — CHANGES_REQUESTED** [review]
Added some review comments. Otherwise some nice cleanups here! 🙌

File: src/index.test.ts [hunk 11]
```diff
@@ -1498,12 +1505,81 @@ describe('Running @erc725/erc725.js tests...', () => {
...
+
+        const decodedPermissions = decodePermissions(encoded);
+        decodedPermissions.CALL = false;
+        const reencodePermissions = encodePermissions(decodedPermissions);
+        const redecodedPermissions = decodePermissions(reencodePermissions);
+
+        assert.strictEqual(
+          redecodedPermissions.CALL,
+          false,
+          'Re-reencoded permissions includes CALL',
```
Comment (CJ42) [inline src/index.test.ts:1551]: isn't this error message incorrect?

```suggestion
          'Re-reencoded permissions does not include CALL',
```

File: src/index.test.ts [hunk 12]
```diff
@@ -1498,12 +1505,81 @@ describe('Running @erc725/erc725.js tests...', () => {
...
+        decodedPermissions.DELEGATECALL = true;
+        const reencodePermissions = encodePermissions(decodedPermissions);
+        const redecodedPermissions = decodePermissions(reencodePermissions);
+
+        assert.strictEqual(
+          redecodedPermissions.SUPER_DELEGATECALL,
+          true,
+          'Re-reencoded permissions includes SUPER_DELEGATECALL',
+        );
+        assert.strictEqual(
```
Comment (CJ42) [inline src/index.test.ts:1570]: I guess we can also add a test for:

```js
        assert.strictEqual(
          redecodedPermissions.ALL_PERMISSIONS,
          true,
          'Re-reencoded permissions includes ALL_PERMISSIONS',
        );
```

File: package.json (hunk 5 above)
Comment (CJ42) [reply package.json:3]: will be overriten by release please, but let's remove it in case.

File: src/index.test.ts (hunk 7 above)
Comment (CJ42) [reply src/index.test.ts:1522]: yes that's the case. All good

**Review by CJ42 — APPROVED** [review]

**Pull request merged**